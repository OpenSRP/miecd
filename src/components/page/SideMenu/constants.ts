import { IconProp } from '@fortawesome/fontawesome-svg-core';
import { NavigationModule, PageLink } from './SubMenu';
import { TFunction } from 'react-i18next';
import {
    ANC_URL,
    CHILD_URL,
    CLIENT_URL,
    HOME_URL,
    HOUSEHOLD_URL,
    LOCATIONS_URL,
    NBC_AND_PNC_ANALYSIS_URL,
    NBC_AND_PNC_COMPARTMENTS_URL,
    NBC_AND_PNC_LOGFACE_URL,
    NBC_AND_PNC_URL,
    NUTRITION_ANALYSIS_URL,
    NUTRITION_COMPARTMENTS_URL,
    NUTRITION_LOGFACE_URL,
    NUTRITION_URL,
    PREGNANCY_ANALYSIS_URL,
    PREGNANCY_COMPARTMENTS_URL,
    PREGNANCY_LOGFACE_URL,
    PREGNANCY_URL,
    REPORTS_URL,
    ROLE_URL,
    TEAM_URL,
    USER_URL,
} from '../../../constants';

import {
    ENABLE_ANC_PAGE,
    ENABLE_CHILD_PAGE,
    ENABLE_CLIENT_PAGE,
    ENABLE_HOUSEHOLD_PAGE,
    ENABLE_LOCATION_PAGE,
    ENABLE_ROLE_PAGE,
    ENABLE_TEAM_PAGE,
    ENABLE_USER_PAGE,
} from '../../../configs/env';
import { Dictionary } from '@onaio/utils';

export const navigationPageLinks = (t: TFunction): Dictionary<PageLink> => {
    return {
        // Page links
        CLIENT_PAGE_NAVIGATION: {
            label: t('All clients'),
            url: CLIENT_URL,
        },
        HOUSEHOLD_PAGE_NAVIGATION: {
            label: t('Household'),
            url: HOUSEHOLD_URL,
        },
        ANC_PAGE_NAVIGATION: {
            label: t('ANC'),
            url: ANC_URL,
        },
        CHILD_PAGE_NAVIGATION: {
            label: t('Child'),
            url: CHILD_URL,
        },

        USERS_PAGE_NAVIGATION: {
            label: t('Users'),
            url: USER_URL,
        },
        ROLES_PAGE_NAVIGATION: {
            label: t('Roles'),
            url: ROLE_URL,
        },
        TEAMS_PAGE_NAVIGATION: {
            label: t('Teams'),
            url: TEAM_URL,
        },
        LOCATIONS_PAGE_NAVIGATION: {
            label: t('Locations'),
            url: LOCATIONS_URL,
        },

        PREGNANCY_PAGE_NAVIGATION: {
            label: t('Pregnancy'),
            url: PREGNANCY_URL,
        },

        NUTRITION_PAGE_NAVIGATION: {
            label: t('Nutrition'),
            url: NUTRITION_URL,
        },

        NEWBORN_AND_POSTNATAL_PAGE_NAVIGATION: {
            label: t('NBC & PNC'),
            url: NBC_AND_PNC_URL,
        },

        PREGNANCY_COMPARTMENTS_PAGE_NAVIGATION: {
            label: t('Compartments'),
            url: PREGNANCY_COMPARTMENTS_URL,
        },

        NBC_AND_PNC_COMPARTMENTS_PAGE_NAVIGATION: {
            label: t('Compartments'),
            url: NBC_AND_PNC_COMPARTMENTS_URL,
        },

        NUTRITION_COMPARTMENTS_PAGE_NAVIGATION: {
            label: t('Compartments'),
            url: NUTRITION_COMPARTMENTS_URL,
        },

        NBC_AND_PNC_LOGFACE_PAGE_NAVIGATION: {
            label: t('Log face'),
            url: NBC_AND_PNC_LOGFACE_URL,
        },

        PREGNANCY_LOGFACE_PAGE_NAVIGATION: {
            label: t('Log face'),
            url: PREGNANCY_LOGFACE_URL,
        },

        NUTRITION_LOGFACE_PAGE_NAVIGATION: {
            label: t('Log face'),
            url: NUTRITION_LOGFACE_URL,
        },

        PREGNANCY_ANALYSIS_PAGE_NAVIGATION: {
            label: t('Analysis'),
            url: PREGNANCY_ANALYSIS_URL,
        },

        NBC_AND_PNC_ANALYSIS_PAGE_NAVIGATION: {
            label: t('Analysis'),
            url: NBC_AND_PNC_ANALYSIS_URL,
        },

        NUTRITION_ANALYSIS_PAGE_NAVIGATION: {
            label: t('Analysis'),
            url: NUTRITION_ANALYSIS_URL,
        },
    };
};

// icons
export const homeNavIcon: IconProp = ['fas', 'home'];
export const pregancyNavIcon: IconProp = ['far', 'user'];
export const nbcandpncNavIcon: IconProp = ['far', 'user'];
export const nutritionNavIcon: IconProp = ['far', 'user'];
export const newbornAndPostnatalNavIcon: IconProp = ['far', 'user'];
export const clientModuleNavIcon: IconProp = ['far', 'user'];
export const reportModuleNavIcon: IconProp = ['fas', 'chart-line'];
export const adminModuleNavIcon: IconProp = ['fas', 'cog'];

// Navigation links for navigation module. collapses when clicked to display the pageLinks
export const modulePageLinks = (t: TFunction) => {
    const links = {
        HOME_PARENT_NAV: {
            icon: homeNavIcon,
            label: t('Home'),
            url: HOME_URL,
        },
        NBC_AND_PNC_PARENT_NAV: {
            icon: nbcandpncNavIcon,
            label: t('NBC & PNC'),
            url: NBC_AND_PNC_URL,
        },
        NUTRITION_PARENT_NAVIGATION: {
            icon: nutritionNavIcon,
            label: t('Nutrition'),
            url: NUTRITION_URL,
        },
        PREGNANCY_MODULE_PARENT_NAV: {
            icon: pregancyNavIcon,
            label: t('Pregnancy'),
            url: PREGNANCY_URL,
        },
        CLIENT_MODULE_PARENT_NAV: {
            icon: clientModuleNavIcon,
            label: t('Client Records'),
        },
        REPORT_MODULE_PARENT_NAV: {
            icon: reportModuleNavIcon,
            label: t('Reports'),
            url: REPORTS_URL,
        },
        ADMIN_MODULE_PARENT_NAV: {
            icon: adminModuleNavIcon,
            label: t('Admin'),
        },
    };
    return links;
};

export const navigationModulesFactory = (t: TFunction) => {
    const {
        PREGNANCY_MODULE_PARENT_NAV,
        NBC_AND_PNC_PARENT_NAV,
        NUTRITION_PARENT_NAVIGATION,
        CLIENT_MODULE_PARENT_NAV,
        HOME_PARENT_NAV,
        ADMIN_MODULE_PARENT_NAV,
        REPORT_MODULE_PARENT_NAV,
    } = modulePageLinks(t);
    const {
        PREGNANCY_COMPARTMENTS_PAGE_NAVIGATION,
        PREGNANCY_ANALYSIS_PAGE_NAVIGATION,
        PREGNANCY_LOGFACE_PAGE_NAVIGATION,
        NBC_AND_PNC_LOGFACE_PAGE_NAVIGATION,
        NBC_AND_PNC_COMPARTMENTS_PAGE_NAVIGATION,
        NBC_AND_PNC_ANALYSIS_PAGE_NAVIGATION,
        NUTRITION_LOGFACE_PAGE_NAVIGATION,
        NUTRITION_COMPARTMENTS_PAGE_NAVIGATION,
        NUTRITION_ANALYSIS_PAGE_NAVIGATION,
        CLIENT_PAGE_NAVIGATION,
        HOUSEHOLD_PAGE_NAVIGATION,
        ANC_PAGE_NAVIGATION,
        CHILD_PAGE_NAVIGATION,
        USERS_PAGE_NAVIGATION,
        ROLES_PAGE_NAVIGATION,
        TEAMS_PAGE_NAVIGATION,
        LOCATIONS_PAGE_NAVIGATION,
    } = navigationPageLinks(t);
    // nav module objects
    const navigationModules: Dictionary<NavigationModule> = {
        HOME_NAVIGATION_MODULE: {
            childNavs: [],
            parentNav: HOME_PARENT_NAV,
        },

        PREGNANCY_NAVIGATION_MODULE: {
            childNavs: [
                PREGNANCY_LOGFACE_PAGE_NAVIGATION,
                PREGNANCY_COMPARTMENTS_PAGE_NAVIGATION,
                PREGNANCY_ANALYSIS_PAGE_NAVIGATION,
            ].filter((childNav): childNav is PageLink => typeof childNav !== 'boolean'),
            parentNav: PREGNANCY_MODULE_PARENT_NAV,
        },

        NBC_AND_PNC_NAVIGATION_MODULE: {
            childNavs: [
                NBC_AND_PNC_LOGFACE_PAGE_NAVIGATION,
                NBC_AND_PNC_COMPARTMENTS_PAGE_NAVIGATION,
                NBC_AND_PNC_ANALYSIS_PAGE_NAVIGATION,
            ],
            parentNav: NBC_AND_PNC_PARENT_NAV,
        },

        NUTRITION_MODULE: {
            childNavs: [
                NUTRITION_LOGFACE_PAGE_NAVIGATION,
                NUTRITION_COMPARTMENTS_PAGE_NAVIGATION,
                NUTRITION_ANALYSIS_PAGE_NAVIGATION,
            ],
            parentNav: NUTRITION_PARENT_NAVIGATION,
        },

        CLIENT_NAVIGATION_MODULE: {
            childNavs: [
                ENABLE_CLIENT_PAGE && CLIENT_PAGE_NAVIGATION,
                ENABLE_HOUSEHOLD_PAGE && HOUSEHOLD_PAGE_NAVIGATION,
                ENABLE_ANC_PAGE && ANC_PAGE_NAVIGATION,
                ENABLE_CHILD_PAGE && CHILD_PAGE_NAVIGATION,
            ].filter((childNav): childNav is PageLink => typeof childNav !== 'boolean'),
            parentNav: CLIENT_MODULE_PARENT_NAV,
        },
        REPORT_NAVIGATION_MODULE: {
            childNavs: [],
            parentNav: REPORT_MODULE_PARENT_NAV,
        },
        ADMIN_NAVIGATION_MODULE: {
            childNavs: [
                ENABLE_USER_PAGE && USERS_PAGE_NAVIGATION,
                ENABLE_ROLE_PAGE && ROLES_PAGE_NAVIGATION,
                ENABLE_TEAM_PAGE && TEAMS_PAGE_NAVIGATION,
                ENABLE_LOCATION_PAGE && LOCATIONS_PAGE_NAVIGATION,
            ].filter<PageLink>((childNav: PageLink | boolean): childNav is PageLink => typeof childNav !== 'boolean'),
            parentNav: ADMIN_MODULE_PARENT_NAV,
        },
    };
    return navigationModules;
};

/** Enable Clients Module from its child pages */
export const ENABLE_CLIENT_RECORDS_MODULE =
    ENABLE_ANC_PAGE || ENABLE_CLIENT_PAGE || ENABLE_HOUSEHOLD_PAGE || ENABLE_HOUSEHOLD_PAGE;
export type ENABLE_CLIENT_RECORDS_MODULE = typeof ENABLE_CLIENT_RECORDS_MODULE;

/** Enable Admin Module from its child pages */
export const ENABLE_ADMIN_MODULE = ENABLE_USER_PAGE || ENABLE_ROLE_PAGE || ENABLE_TEAM_PAGE || ENABLE_LOCATION_PAGE;
export type ENABLE_ADMIN_MODULE = typeof ENABLE_ADMIN_MODULE;
