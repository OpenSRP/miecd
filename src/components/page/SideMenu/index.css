.side-menu-container {
    width: 300px;
    height: 100%;
    background: #f8fafe;
    border-right: 1px solid #dfe3e8;
}

.side-menu-container ul.rc-menu {
    box-shadow: none;
    border: none;
    text-decoration: none;
}

.side-menu-container ul.rc-menu a {
    display: inline-block;
    width: 100%;
    color: rgb(102, 102, 102);
}
.side-menu-container .rc-menu-item-active:not(.rc-menu-item-selected) {
    background: transparent !important;
}

.side-menu-container .rc-menu-item-active a {
    color: #0093db !important;
    text-decoration: none;
}

.side-menu-container .rc-menu-item-selected {
    border-right: 5px solid #0093db;
    background: rgba(1, 146, 212, 0.1);
    color: #0093db;
}

.side-menu-container .rc-menu-item-selected a {
    color: #0093db !important;
}

.rc-menu-submenu-selected {
    background: transparent !important;
    color: #0093db !important;
}
