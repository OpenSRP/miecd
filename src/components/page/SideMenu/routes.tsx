import { IconProp } from '@fortawesome/fontawesome-svg-core';
import { TFunction } from 'react-i18next';
import {
    HOME_URL,
    NBC_AND_PNC_ANALYSIS_URL,
    NBC_AND_PNC_COMPARTMENTS_URL,
    NBC_AND_PNC_LOGFACE_URL,
    NUTRITION_ANALYSIS_URL,
    NUTRITION_COMPARTMENTS_URL,
    NUTRITION_LOGFACE_URL,
    PREGNANCY_ANALYSIS_URL,
    PREGNANCY_COMPARTMENTS_URL,
    PREGNANCY_LOGFACE_URL,
    URL_LOCATION_UNIT,
    URL_LOCATION_UNIT_GROUP,
    URL_TEAMS,
    URL_TEAM_ASSIGNMENT,
    URL_USER,
    URL_USER_GROUPS,
    URL_USER_ROLES,
} from '../../../constants';

import {
    ENABLE_LOCATIONS,
    ENABLE_TEAMS,
    ENABLE_TEAMS_ASSIGNMENT_MODULE,
    ENABLE_USERS,
    OPENSRP_ROLES,
} from '../../../configs/env';
import { isAuthorized } from '@opensrp/react-utils';
import { ReactComponent as HomeLogo } from '../../../assets/menuIcons/home.svg';
import { ReactComponent as NbcAndPncLogo } from '../../../assets/menuIcons/nbcandpnc.svg';
import { ReactComponent as NutritionLogo } from '../../../assets/menuIcons/nutrition.svg';
import { ReactComponent as PregnancyLogo } from '../../../assets/menuIcons/pregnancy.svg';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { ReactNode } from 'react';

/** Interface for menu items */
export interface Route {
    key: string;
    enabled?: boolean;
    url?: string;
    title: string;
    otherProps?: {
        icon?: ReactNode;
    };
    children?: Route[];
}

// icons
export const adminModuleNavIconString: IconProp = ['fas', 'cog'];
export const adminModuleNavIcon = <FontAwesomeIcon icon={adminModuleNavIconString} />;

/** Gets Routes For Application
 *
 * @param roles User's roles
 * @returns {Route[]} returns generated routes
 */
export function getRoutes(roles: string[], t: TFunction): Route[] {
    const activeRoles = OPENSRP_ROLES;

    const routes: Route[] = [
        {
            otherProps: { icon: <HomeLogo /> },
            title: t('Home'),
            key: 'home',
            enabled: true,
            url: HOME_URL,
        },
        {
            otherProps: { icon: <PregnancyLogo /> },
            title: t('Pregnancy'),
            key: 'pregnancy',
            enabled: true,
            children: [
                {
                    title: t('Log face'),
                    key: 'pregnancyLogFace',
                    url: PREGNANCY_LOGFACE_URL,
                    enabled: true,
                },
                {
                    title: t('Compartments'),
                    key: 'pregnancyCompartments',
                    url: PREGNANCY_COMPARTMENTS_URL,
                    enabled: true,
                },
                {
                    title: t('Analysis'),
                    key: 'pregnancyAnalysis',
                    url: PREGNANCY_ANALYSIS_URL,
                    enabled: true,
                },
            ],
        },
        {
            otherProps: { icon: <NbcAndPncLogo /> },
            title: t('NBC & PNC'),
            key: 'nbcPnc',
            enabled: true,
            children: [
                {
                    title: t('Log face'),
                    key: 'nbcPncLogFace',
                    url: NBC_AND_PNC_LOGFACE_URL,
                    enabled: true,
                },
                {
                    title: t('Compartments'),
                    key: 'nbcPncCompartments',
                    url: NBC_AND_PNC_COMPARTMENTS_URL,
                    enabled: true,
                },
                {
                    title: t('Analysis'),
                    key: 'nbcPncAnalysis',
                    url: NBC_AND_PNC_ANALYSIS_URL,
                    enabled: true,
                },
            ],
        },
        {
            otherProps: { icon: <NutritionLogo /> },
            title: t('Nutrition'),
            key: 'nutrition',
            enabled: true,
            children: [
                {
                    title: t('Log face'),
                    key: 'nutritionLogFace',
                    url: NUTRITION_LOGFACE_URL,
                    enabled: true,
                },
                {
                    title: t('Compartments'),
                    key: 'nutritionCompartments',
                    url: NUTRITION_COMPARTMENTS_URL,
                    enabled: true,
                },
                {
                    title: t('Analysis'),
                    key: 'nutritionAnalysis',
                    url: NUTRITION_ANALYSIS_URL,
                    enabled: true,
                },
            ],
        },
        {
            otherProps: { icon: adminModuleNavIcon },
            title: t('Admin'),
            key: 'admin',
            enabled:
                (ENABLE_LOCATIONS || ENABLE_TEAMS || ENABLE_USERS) &&
                roles &&
                isAuthorized(roles, activeRoles.ADMIN?.split(',') ?? []),
            url: '/admin',
            children: [
                {
                    title: t('User Management'),
                    key: 'users',
                    enabled: ENABLE_USERS && roles && isAuthorized(roles, activeRoles.USERS?.split(',') ?? []),
                    children: [
                        { title: t('Users'), key: 'user', url: URL_USER, enabled: true },
                        { title: t('User groups'), key: 'user-groups', url: URL_USER_GROUPS, enabled: true },
                        { title: t('User roles'), key: 'user-roles', url: URL_USER_ROLES, enabled: true },
                    ],
                },
                {
                    title: 'Locations',
                    key: 'locations',
                    enabled: ENABLE_LOCATIONS && roles && isAuthorized(roles, activeRoles.LOCATIONS?.split(',') ?? []),
                    children: [
                        { title: t('Location unit'), url: URL_LOCATION_UNIT, key: 'location-unit', enabled: true },
                        {
                            title: t('Location unit group'),
                            url: URL_LOCATION_UNIT_GROUP,
                            enabled: true,
                            key: 'location-group',
                        },
                    ],
                },
                {
                    title: t('Team management'),
                    key: 'teams',
                    enabled: ENABLE_TEAMS && roles && isAuthorized(roles, activeRoles.TEAMS?.split(',') ?? []),
                    children: [
                        { title: t('Teams'), url: URL_TEAMS, key: 'teams-list', enabled: true },
                        {
                            title: t('Teams assignment'),
                            url: URL_TEAM_ASSIGNMENT,
                            key: 'team-assignment',
                            enabled: ENABLE_TEAMS_ASSIGNMENT_MODULE,
                        },
                    ],
                },
            ],
        },
    ];

    return filterFalsyRoutes(routes);
}

/** Removes the disabled Routes from
 *
 * @param routes all routes
 * @returns {Route[]} returns only enabled routes
 */
export function filterFalsyRoutes(routes: Route[]): Route[] {
    return routes
        .filter((e) => e.enabled)
        .map((e) => {
            return e.children ? { ...e, children: filterFalsyRoutes(e.children) } : e;
        });
}

/** Gets Active key for menu based on path from routes
 *
 * @param routes - The routes to get search the active route from
 * @param path - an array of menu location path
 */
export function getActiveKey(path: string, routes: Route[]) {
    let activeKey: string | undefined;

    function mapMenus(menu: Route) {
        // Matching Url
        if (menu.url && path.includes(menu.url)) activeKey = menu.key;

        // Exact Match
        if (path === menu.url) activeKey = menu.key;
        // Trying to Match with Children
        else if (menu.children && path !== menu.url) menu.children.forEach(mapMenus);
    }
    routes.forEach(mapMenus);

    return activeKey;
}
