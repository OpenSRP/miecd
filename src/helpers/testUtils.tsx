import { mount, MountRendererProps } from 'enzyme';
import React from 'react';
import { I18nextProvider } from 'react-i18next';
import i18n from '../mls';
import { act } from 'react-dom/test-utils';

export const mountWithTranslations: typeof mount = (component: React.ReactNode, options?: MountRendererProps) => {
    return mount(<I18nextProvider i18n={i18n}>{component}</I18nextProvider>, options);
};

/**
 * Utility to await promises to resolve in tests
 * @param ms time in ms
 */
export async function waitForPromises() {
    await act(async () => {
        await new Promise((res) => setImmediate(res));
    });
}
