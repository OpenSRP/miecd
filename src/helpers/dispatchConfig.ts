import { ConfigState, setAllConfigs, getAllConfigs } from '@opensrp/pkg-config';
import { OPENSRP_REST_API_BASE_URL } from 'configs/settings';
import { BACKEND_ACTIVE } from '../configs/env';
import { BACKEND_LOGIN_URL, REACT_LOGIN_URL } from '../constants';
import i18n from '../mls';

export const APP_LOGIN_URL = BACKEND_ACTIVE ? BACKEND_LOGIN_URL : REACT_LOGIN_URL;

const defaultvalues = getAllConfigs();

const configObject: ConfigState = {
    ...defaultvalues,
    appLoginURL: APP_LOGIN_URL,
    opensrpBaseURL: OPENSRP_REST_API_BASE_URL,
    i18n,
};

setAllConfigs(configObject);
