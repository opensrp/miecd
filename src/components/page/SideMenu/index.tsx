import React from 'react';
import { withRouter } from 'react-router';
import { Link, useLocation } from 'react-router-dom';
import { useTranslation } from 'react-i18next';
import Menu from 'rc-menu';
import './index.css';
import { getActiveKey, getRoutes, Route } from './routes';
import { getExtraData } from '@opensrp/store';
import 'rc-menu/assets/index.css';
import { useSelector } from 'react-redux';

/** The Sidebar component */
export const SidebarComponent: React.FC = () => {
    const { t } = useTranslation();
    const location = useLocation();
    const [openKeys, setOpenKeys] = React.useState<React.Key[]>([]);

    const extraData = useSelector((state) => getExtraData(state));
    const { roles } = extraData;

    const routes = React.useMemo(() => getRoutes(roles as string[], t), [roles, t]);

    const sidebaritems: JSX.Element[] = React.useMemo(() => {
        function mapChildren(route: Route) {
            const icon = route.otherProps?.icon;
            if (route.children) {
                const title = (
                    <>
                        {icon}&nbsp;&nbsp;&nbsp;{route.title}
                    </>
                );

                return (
                    <Menu.SubMenu key={route.key} title={title}>
                        {route.children.map(mapChildren)}
                    </Menu.SubMenu>
                );
            } else if (route.url) {
                return (
                    <Menu.Item key={route.key}>
                        <Link className="admin-link" to={route.url}>
                            {icon ? (
                                <>
                                    {icon}&nbsp;&nbsp;&nbsp;{route.title}
                                </>
                            ) : (
                                route.title
                            )}
                        </Link>
                    </Menu.Item>
                );
            } else {
                return <Menu.Item key={route.key}>{route.title}</Menu.Item>;
            }
        }

        return routes.map(mapChildren);
    }, [routes]);

    const activeLocationPaths = location.pathname.split('/').filter((locString: string) => locString.length);
    const activeKey = getActiveKey(location.pathname, routes);

    return (
        <div>
            <div className="side-menu-container">
                <Menu
                    key="main-menu"
                    selectedKeys={[activeKey ?? '']}
                    openKeys={openKeys.length ? (openKeys as string[]) : activeLocationPaths}
                    defaultOpenKeys={activeLocationPaths}
                    defaultSelectedKeys={[activeKey ?? '']}
                    onOpenChange={(keys) => setOpenKeys(keys)}
                    mode="inline"
                    className="menu-dark"
                >
                    {sidebaritems}
                </Menu>
            </div>
        </div>
    );
};

const Sidebar = withRouter(SidebarComponent);

export default Sidebar;
