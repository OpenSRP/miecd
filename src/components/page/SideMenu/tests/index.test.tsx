import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { createBrowserHistory } from 'history';
import React from 'react';
import { Router } from 'react-router';
import SideMenu from '..';
import { PREGNANCY } from '../../../../constants';
import { mountWithTranslations } from '../../../../helpers/testUtils';

const history = createBrowserHistory();

jest.mock('../../../../configs/env');

describe('components/page/SideMenu', () => {
    beforeEach(() => {
        jest.resetAllMocks();
    });

    // eslint-disable-next-line jest/expect-expect
    it('renders without crashing', () => {
        shallow(
            <Router history={history}>
                <SideMenu authenticated />
            </Router>,
        );
    });

    it('renders side menu correctly', () => {
        const wrapper = mountWithTranslations(
            <Router history={history}>
                <SideMenu authenticated />
            </Router>,
        );
        /** client Collapse SubMenu renders correctly */
        expect(toJson(wrapper.find('Nav .side-collapse-nav').first())).toMatchSnapshot();
        wrapper.unmount();
    });

    it('manages state correctly', () => {
        const wrapper = mountWithTranslations(
            <Router history={history}>
                <SideMenu authenticated />
            </Router>,
        );

        expect(wrapper.find('SideMenu').state('collapsedModuleLabel')).toEqual('');

        wrapper.find('ul#Pregnancy').simulate('click');
        expect(wrapper.find('SideMenu').state('collapsedModuleLabel')).toEqual(PREGNANCY);
        wrapper.unmount();
    });

    it('sets the collapsedModuleLabel correctly from clicks on parentNavs', async () => {
        // clicking changes the collapsedModuleLabel state and collapses
        // nav to reveal child navigation

        const wrapper = mountWithTranslations(
            <Router history={history}>
                <SideMenu authenticated />)
            </Router>,
        );

        // how many parent navigationModules are initially collapsed
        expect(wrapper.find('div.collapse')).toHaveLength(7);
        expect(wrapper.find('div.collapse.show')).toHaveLength(0);
        expect(wrapper.find('SubMenu').at(1).find('Collapse').prop('isOpen')).toBeFalsy();

        // clicking on a parent nav changes the collapsedState for that navigation module
        const pregnancyNav = wrapper.find('div#sub-menu-Pregnancy Nav.side-collapse-nav');
        console.log(toJson(pregnancyNav));
        expect(pregnancyNav.length).toEqual(1);
        pregnancyNav.simulate('click');
        wrapper.update();

        wrapper.find('SubMenu').forEach((div) => expect(expect(div.text()).toMatchSnapshot('sideMenu link labels')));

        expect(wrapper.find('SubMenu').at(1).prop('collapsedModuleLabel')).toEqual('Pregnancy');

        expect(wrapper.find('SubMenu').at(1).find('Collapse').prop('isOpen')).toBeTruthy();

        wrapper.unmount();
    });
});
