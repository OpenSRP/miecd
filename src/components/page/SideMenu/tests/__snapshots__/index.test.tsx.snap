// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/page/SideMenu renders side menu correctly 1`] = `
<ul
  className="side-collapse-nav nav"
  id="Home"
  onClick={[Function]}
>
  <NavItem
    tag="li"
  >
    <li
      className="nav-item"
    >
      <NavLink
        className="nav nav-link side-nav-item active-collapse-menu"
        to="/"
      >
        <Link
          aria-current="page"
          className="nav nav-link side-nav-item active-collapse-menu active"
          style={Object {}}
          to={
            Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": null,
            }
          }
        >
          <LinkAnchor
            aria-current="page"
            className="nav nav-link side-nav-item active-collapse-menu active"
            href="/"
            navigate={[Function]}
            style={Object {}}
          >
            <a
              aria-current="page"
              className="nav nav-link side-nav-item active-collapse-menu active"
              href="/"
              onClick={[Function]}
              style={Object {}}
            >
              <ForwardRef(SvgHome)>
                <svg>
                  home.svg
                </svg>
              </ForwardRef(SvgHome)>
              <span
                className="collapse-menu-title"
              >
                 
                Home
                 
              </span>
            </a>
          </LinkAnchor>
        </Link>
      </NavLink>
    </li>
  </NavItem>
</ul>
`;

exports[`components/page/SideMenu sets the collapsedModuleLabel correctly from clicks on parentNavs: Everything 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -78653,11 +78653,11 @@
                        <div
                          className=\\"side-menu-extend col\\"
                        >
                          <withRouter(SubMenu)
                            childNavs={Array []}
-                           collapsedModuleLabel=\\"\\"
+                           collapsedModuleLabel=\\"Pregnancy\\"
                            customIcon={
                              Object {
                                \\"$$typeof\\": Symbol(react.forward_ref),
                                \\"render\\": [Function SvgHome],
                              }
@@ -78675,11 +78675,11 @@
                            }
                            setCollapsedModuleLabel={[Function anonymous]}
                          >
                            <SubMenu
                              childNavs={Array []}
-                             collapsedModuleLabel=\\"\\"
+                             collapsedModuleLabel=\\"Pregnancy\\"
                              customIcon={
                                Object {
                                  \\"$$typeof\\": Symbol(react.forward_ref),
                                  \\"render\\": [Function SvgHome],
                                }
@@ -78860,11 +78860,11 @@
                                  \\"label\\": \\"Analysis\\",
                                  \\"url\\": \\"/pregnancy_analysis\\",
                                },
                              ]
                            }
-                           collapsedModuleLabel=\\"\\"
+                           collapsedModuleLabel=\\"Pregnancy\\"
                            customIcon={
                              Object {
                                \\"$$typeof\\": Symbol(react.forward_ref),
                                \\"render\\": [Function SvgPregnancy],
                              }
@@ -78897,11 +78897,11 @@
                                    \\"label\\": \\"Analysis\\",
                                    \\"url\\": \\"/pregnancy_analysis\\",
                                  },
                                ]
                              }
-                             collapsedModuleLabel=\\"\\"
+                             collapsedModuleLabel=\\"Pregnancy\\"
                              customIcon={
                                Object {
                                  \\"$$typeof\\": Symbol(react.forward_ref),
                                  \\"render\\": [Function SvgPregnancy],
                                }
@@ -79026,11 +79026,11 @@
                                <Collapse
                                  appear={false}
                                  enter={true}
                                  exit={true}
                                  in={false}
-                                 isOpen={false}
+                                 isOpen={true}
                                  mountOnEnter={false}
                                  onEnter={[Function noop]}
                                  onEntered={[Function noop]}
                                  onEntering={[Function noop]}
                                  onExit={[Function noop]}
@@ -79042,11 +79042,11 @@
                                >
                                  <Transition
                                    appear={false}
                                    enter={true}
                                    exit={true}
-                                   in={false}
+                                   in={true}
                                    mountOnEnter={false}
                                    onEnter={[Function noop]}
                                    onEntered={[Function bound onEntered]}
                                    onEntering={[Function bound onEntering]}
                                    onExit={[Function bound onExit]}
@@ -79054,12 +79054,16 @@
                                    onExiting={[Function bound onExiting]}
                                    timeout={350}
                                    unmountOnExit={false}
                                  >
                                    <div
-                                     className=\\"collapse\\"
-                                     style={Object {}}
+                                     className=\\"collapsing\\"
+                                     style={
+                                       Object {
+                                         \\"height\\": 0,
+                                       }
+                                     }
                                    >
                                      <Nav
                                        key=\\"child-nav-0\\"
                                        tag=\\"ul\\"
                                        vertical={false}
@@ -79251,11 +79255,11 @@
                                  \\"label\\": \\"Analysis\\",
                                  \\"url\\": \\"/nbc_and_pnc_analysis\\",
                                },
                              ]
                            }
-                           collapsedModuleLabel=\\"\\"
+                           collapsedModuleLabel=\\"Pregnancy\\"
                            customIcon={
                              Object {
                                \\"$$typeof\\": Symbol(react.forward_ref),
                                \\"render\\": [Function SvgNbcandpnc],
                              }
@@ -79288,11 +79292,11 @@
                                    \\"label\\": \\"Analysis\\",
                                    \\"url\\": \\"/nbc_and_pnc_analysis\\",
                                  },
                                ]
                              }
-                             collapsedModuleLabel=\\"\\"
+                             collapsedModuleLabel=\\"Pregnancy\\"
                              customIcon={
                                Object {
                                  \\"$$typeof\\": Symbol(react.forward_ref),
                                  \\"render\\": [Function SvgNbcandpnc],
                                }
@@ -79642,11 +79646,11 @@
                                  \\"label\\": \\"Analysis\\",
                                  \\"url\\": \\"/nutrition_analysis\\",
                                },
                              ]
                            }
-                           collapsedModuleLabel=\\"\\"
+                           collapsedModuleLabel=\\"Pregnancy\\"
                            customIcon={
                              Object {
                                \\"$$typeof\\": Symbol(react.forward_ref),
                                \\"render\\": [Function SvgNutrition],
                              }
@@ -79679,11 +79683,11 @@
                                    \\"label\\": \\"Analysis\\",
                                    \\"url\\": \\"/nutrition_analysis\\",
                                  },
                                ]
                              }
-                             collapsedModuleLabel=\\"\\"
+                             collapsedModuleLabel=\\"Pregnancy\\"
                              customIcon={
                                Object {
                                  \\"$$typeof\\": Symbol(react.forward_ref),
                                  \\"render\\": [Function SvgNutrition],
                                }"
`;
